presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -586,4 +596,165 @@ public void prepareFinal()
             });
         }
     }
+
+    /**
+     * SpillableFinalOnlyGroupedAccumulator enables spilling for FinalOnlyGroupedAccumulators
+     */
+    private static class SpillableFinalOnlyGroupedAccumulator
+            implements GroupedAccumulator
+    {
+        private final FinalOnlyGroupedAccumulator delegate;
+        private final List<Type> spillingTypes;
+        private List<GroupIdPage> rawInputs = new ArrayList<>();
+        private ObjectBigArray<RowBlockBuilder> blockBuilders;
+        private long rawInputsSize;
+
+        public SpillableFinalOnlyGroupedAccumulator(List<Type> types, FinalOnlyGroupedAccumulator delegate)
+        {
+            this.delegate = delegate;
+            this.spillingTypes = types;
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return delegate.getEstimatedSize() + rawInputsSize + (blockBuilders == null ? 0 : blockBuilders.sizeOf());
+        }
+
+        @Override
+        public Type getFinalType()
+        {
+            return delegate.getFinalType();
+        }
+
+        @Override
+        public Type getIntermediateType()
+        {
+            return new ArrayType(RowType.anonymous(spillingTypes));
+        }
+
+        @Override
+        public void addInput(GroupByIdBlock groupIdsBlock, Page page)
+        {
+            checkState(rawInputs != null && blockBuilders == null);
+            rawInputsSize += page.getSizeInBytes();
+            rawInputs.add(new GroupIdPage(groupIdsBlock, page));
+        }
+
+        @Override
+        public void addIntermediate(GroupByIdBlock groupIdsBlock, Block block)
+        {
+            checkState(rawInputs != null && blockBuilders == null);
+            checkState(block instanceof ArrayBlock);
+            ArrayBlock arrayBlock = (ArrayBlock) block;
+
+            ColumnarArray columnarArray = toColumnarArray(block);
+            ColumnarRow columnarRow = toColumnarRow(columnarArray.getElementsBlock());
+            int newPositionCount = columnarRow.getPositionCount();
+            long[] newGroupIds = new long[newPositionCount];
+            boolean[] nulls = new boolean[newPositionCount];
+            int currentRowBlockIndex = 0;
+            for (int groupIdPosition = 0; groupIdPosition < groupIdsBlock.getPositionCount(); groupIdPosition++) {
+                for (int unused = 0; unused < arrayBlock.getBlock(groupIdPosition).getPositionCount(); unused++) {
+                    newGroupIds[currentRowBlockIndex] = groupIdsBlock.getGroupId(groupIdPosition);
+                    nulls[currentRowBlockIndex] = groupIdsBlock.isNull(groupIdPosition);
+                    currentRowBlockIndex++;
+                }
+            }
+
+            Block[] blocks = new Block[spillingTypes.size()];
+            for (int channel = 0; channel < spillingTypes.size(); channel++) {
+                blocks[channel] = columnarRow.getField(channel);
+            }
```

Comment:
It took me a long while to understand what this part is doing. We need detailed comments to explain:
- `ColumnarArray` flattens the squashed arrays; so there is no need to flatten `block` again.
- `newPositionCount` is the unrolled position count already.
- `unused` is unused because we are unrolling for the same group id; this corresponds to how `ColumnarArray` unrolls internal block.
- `ColumnarRow` already contains the unrolled blocks.